Add built-in Migration Assistant field data type transformation documentation (#10649)
opensearch-project/documentation-website: _migration-assistant/migration-phases/migrate-metadata/transform-string-text-keyword.md

---
layout: default
title: Transform string to text/keyword
nav_order: 4
parent: Migrate metadata
grand_parent: Migration phases
permalink: /migration-assistant/migration-phases/migrate-metadata/transform-string-text-keyword/
---

# Transform string to text/keyword

Convert field data type string to text/keyword based on field data mappings.

This guide explains how the Migration Assistant automatically handles the deprecated `string` field type during migration from older Elasticsearch versions.

> ## Overview

Reviewer: Minor suggestion : 

In the overview section of these 3 new pages, can we also answer customer's question "how do I know if my data contains such field type".

```
To determine if your Elasticsearch cluster contains `dense_vector` fields that will be transformed during migration, use the mapping API:

- GET /_mapping?filter_path=.mappings.properties..type
- GET /your-index/_mapping/field/*

If you see `"type": "dense_vector"` in the mappings, these fields will be automatically converted to `knn_vector` during migration.
```

      "type": "string",
      "analyzer": "standard",
      "fielddata": true
    },
    "description": {
      "type": "string",
      "index": "analyzed",
      "term_vector": "with_positions"
    }
  }
}</code></pre>
      </td>
      <td style="border: 1px solid #ddd; padding: 8px;">
        <pre><code>{
  "properties": {
    "title": {
      "type": "text",
      "analyzer": "standard",
      "fielddata": true
    },
    "description": {
      "type": "text",
      "term_vector": "with_positions"
    }
  }
}</code></pre>
      </td>
    </tr>
  </tbody>
</table>

## Behavior differences

The Migration Assistant automatically converts all `string` fields during metadata migration. No additional configuration is required.

### Mixed field types

If your source cluster already contains a mix of `string`, `text`, and `keyword` fields (common in Elasticsearch 5.x with restored Elasticsearch 2.x indices), only the `string` fields will be converted. Existing `text` and `keyword` fields remain unchanged.

### Query and aggregation impacts

After migration, be aware of these key differences:

- **Term queries**: Work differently on `text` (analyzed) vs `keyword` (exact match) fields

> - **Aggregations**: `text` fields cannot be used for aggregations unless fielddata is enabled; `keyword` fields are aggregation-ready

Reviewer: Vale is upset. Try using `fielddata`

---
layout: default
title: Transform string to text/keyword
nav_order: 4
parent: Migrate metadata
grand_parent: Migration phases
permalink: /migration-assistant/migration-phases/migrate-metadata/transform-string-text-keyword/
---

# Transform string to text/keyword fields

> This guide explains how the Migration Assistant automatically handles the deprecated `string` field type during migration from older Elasticsearch versions.

Reviewer 2: ```suggestion
This guide explains how Migration Assistant automatically handles the deprecated `string` field type during migration from earlier Elasticsearch versions.
```

## Overview

The `string` field type was the primary text field type in early versions of Elasticsearch but was deprecated in Elasticsearch 5.x and completely removed in Elasticsearch 6.0. In Elasticsearch 5.x, it remained available in backward compatibility mode only.

> When migrating from Elasticsearch 1.x through 5.x to newer versions, the Migration Assistant automatically converts `string` field types to their modern equivalents: `text` for analyzed fields and `keyword` for non-analyzed fields.

Reviewer 2: ```suggestion
When migrating from Elasticsearch 1.x through 5.x to later versions, Migration Assistant automatically converts `string` field types to their modern equivalents: `text` for analyzed fields and `keyword` for non-analyzed fields.
```

> To determine if an Elasticsearch cluster uses `string` field types, make a call to your source cluster's `GET /_mapping` api. On the Migration Console, run `console clusters curl source_cluster "/_mapping"`.  If you see `"type":"string"`, then this transformation is applicable and these fields will be automatically transformed during migration.

Reviewer 2: ```suggestion
To determine whether an Elasticsearch cluster uses `string` field types, make a call to your source cluster's `GET /_mapping` API. In the migration console, run `console clusters curl source_cluster "/_mapping"`. If you see `"type":"string"`, then this transformation is applicable and these fields will be automatically transformed during migration.
```

## Compatibility

The `string` to `text`/`keyword` field type transformation applies to:

> - **Source clusters**: Elasticsearch 1.x - 5.x

Reviewer 2: ```suggestion
- **Source clusters**: Elasticsearch 1.x--5.x
```

- **Target clusters**: Elasticsearch 5.x+ or OpenSearch 1.x+

> - **Automatic conversion**: No configuration required during metadata

Reviewer 2: A word appears to be missing here.

## Automatic conversion logic

> The Migration Assistant determines whether to convert a `string` field to `text` or `keyword` based on the field's `index` property.

Reviewer 2: ```suggestion
Migration Assistant determines whether to convert a `string` field to `text` or `keyword` based on the field's `index` property.
```

### Conversion to `keyword`
Fields are converted to `keyword` when:
- `index: "not_analyzed"`
- `index: "no"`
- `index: false`

### Conversion to `text`
Fields are converted to `text` when:

> - `index: "analyzed"` (default behavior)

Reviewer 2: ```suggestion
- `index: "analyzed"` (default behavior).
```

> - `index` property is not specified

Reviewer 2: ```suggestion
- The `index` property is not specified.
```

> - Any other `index` value

Reviewer 2: This line should be formatted in the same way as the preceding two. This also appears to be an incomplete phrase.

### Property cleanup

> During conversion, the Migration Assistant also cleans up properties that are incompatible with the target field type:

Reviewer 2: ```suggestion
During conversion, Migration Assistant also cleans up properties that are incompatible with the target field type.
```

**For `keyword` fields**, the following properties are removed:
- `analyzer`
- `search_analyzer`
- `position_increment_gap`
- `term_vector`
- `fielddata`

**For `text` fields**, the following properties are removed:
- `doc_values`
- `null_value` (if present)

Additionally, legacy `index` values are normalized:

> - `index: "analyzed"` and `index: "not_analyzed"` are removed (ES 5.x default is `true`)

Reviewer 2: ```suggestion
- `index: "analyzed"` and `index: "not_analyzed"` are removed (Elasticsearch 5.x default is `true`).
```

> - `index: "no"` becomes `index: false`

Reviewer 2: ```suggestion
- `index: "no"` becomes `index: false`.
```

Review your application code for:
- Queries that expect the old `string` field behavior
- Aggregations or sorting on fields now converted to `text`
- Case-sensitive searches on fields now using `keyword`

For fields requiring both analyzed and exact-match capabilities, consider using multi-fields with both `text` and `keyword` mappings after migration.

## Troubleshooting

If you encounter issues with string field conversion:

1. **Verify source version**: Ensure your source cluster is Elasticsearch 1.x through 5.x.

2. **Check migration logs**: Review the detailed migration logs for any warnings or errors:
   ```bash
   tail /shared-logs-output/migration-console-default/*/metadata/*.log
   ```

3. **Validate mappings**: After migration, verify that the field types have been correctly converted:
   ```bash
   GET /your-index/_mapping
   ```

4. **Review field usage**: Ensure that your application queries are compatible with the new field types. Queries that worked with `string` fields should continue to work with `text` and `keyword` fields, but some advanced features may behave differently.

## Related documentation

For additional help with field type transformations, see:

> - [Transform field types]({{site.url}}{{site.baseurl}}/migration-assistant/migration-phases/migrate-metadata/handling-field-type-breaking-changes/) - Configure custom field type transformations

Author: Added
